### DISCUSSION TITLE
Serverside Error Handling (app directory)

### DISCUSSION THREAD

**Discussion by unleashit** [vercel/next.js#46143] [Help]
Just wondering if there's a way to add a global error handler or modify the existing. Right now I'm using Pino which works fine for general logging, but anything thrown and caught by Next's handler are standard console.x. It would also be nice to be able to send a custom status code and message that gets sent to the browser (right now in prod, it sends a generic message with a hash of the server error).

For Next 12 (and I assume 13) with the pages directory, Pino recommends creating a custom server to solve these issues. Will anything change with the app directory?

I tried using an Error component (`error.tsx` in the root). This works on the client, but I don't get any output on the server. For me, the docs are confusing about this. At https://beta.nextjs.org/docs/routing/error-handling#handling-server-errors it says:

> If an error is thrown during [data fetching](https://beta.nextjs.org/docs/data-fetching/fetching) or inside a Server Component, Next.js will forward the resulting Error object to the nearest error.js file as the error prop.

But this component is apparently not called on the server at all, since a plain console.log (in the function body, not a useEffect) only prints on the client.

**Comment by unleashit** [comment]
So since no one has responded and after noticing similar with other users' queries, I can only come to the conclusion that we are left without solution for this.

In case anyone is interested in a solution that involves monkey patching the default error handler, among other things, I did just run into an interesting post: https://jake.tl/notes/2021-04-04-nextjs-preload-hack. It was made at the time of Next 12, so it may or may not work out of the box.

Next.js is cool don't get me wrong, but I've spent so much time trying to find ways around its narrow minded decisions. It's a shame that the trend in front end is to be so opinionated, _before it even has solutions for many things_. There are alternatives of course but they're mostly single maintainers, very old and/or depend on Webpack.

**Reply by XEngine** [reply]
have you find a solution for this case? I am still searching for a viable solution for handling the errors that occurs in the server side not clientside. And yes it looks like we are left without solution for this.

**Reply by unleashit** [reply]
Not sure if there's a "good" answer other than falling in line. Your other options may be:

- Try to use a custom server and define your own error handler. I've never tried this, so not sure if it would help since I assume Next's handler would have probably already sent out the response at a lower level. But if you can somehow get it inserted between what Next is doing and Page.ts it could work.

- Convince Next/Vercel to balance its interest which is currently biased towards newer devs (and all their SMB clients) to also include those who want (or need) more control. Particularly those who host on traditional servers and don't care about the issues of serverless platforms.

I personally believe they have a duty to do the later, since they (and Remix) have done such a great job of convincing the tech industry that a framework is necessary. We plebs have to learn and use it weather we like it or not :kissing_heart:.
